File: EmployeeImportance.py
"""
# Definition for Employee.
class Employee:
    def __init__(self, id: int, importance: int, subordinates: List[int]):
        self.id = id
        self.importance = importance
        self.subordinates = subordinates
"""
# // Time Complexity : O(N)
# // Space Complexity : O(N)

class Solution:
    def getImportance(self, employees: List['Employee'], id: int) -> int:
        employeeMap = collections.defaultdict()
        for e in employees:
            employeeMap[e.id] = e
        return self.dfs(employeeMap, id)

    def dfs(self, employeeMap, id):
        result = employeeMap[id].importance
        for e in employeeMap[id].subordinates:
            result += self.dfs(employeeMap, e)
        return result

File: Employee-Importance.java
/*
// Definition for Employee.
class Employee {
    public int id;
    public int importance;
    public List<Integer> subordinates;
};

Time Complexity : O(n)
Space Complexity : O(n)
*/

class Solution {
    HashMap<Integer, Employee> map = new HashMap<>();
    int total = 0;
    public int getImportance(List<Employee> employees, int id) {
        if(employees == null || employees.size() == 0) return 0;
        
        for(Employee e : employees){
            map.put(e.id, e);
        }
        dfs(id);
        return total;
    }
    
    private void dfs(int id){
        Employee employee = map.get(id);
        total = total + employee.importance;
        for(Integer subid : employee.subordinates){
            dfs(subid);
        } 
    }
}